### ISSUE TITLE
Request Validator name must be unique

### ISSUE THREAD

**Issue by zoellner** [serverless/serverless#10099]
### Are you certain it's a bug?

- [X] Yes, it looks like a bug

### Is the issue caused by a plugin?

- [X] It is not a plugin issue

### Are you using the latest version?

- [X] Yes, I'm using the latest version

### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

### Issue description

I started to see this error in our production build when updating from v2.47.0 to v2.61.0 (since also tried v2.62.0)

```
Serverless Error ----------------------------------------

 An error occurred: ApiGatewayBcbserverapiRequestValidator - Resource handler returned message: "Request Validator name must be unique (Service: ApiGateway, Status Code: 400, Request ID: 30e29e25-296a-42c0-8e6d-cc8a6f31b826, Extended Request ID: null)" (RequestToken: f3479759-3512-b778-ed37-ea6faf4034d5, HandlerErrorCode: AlreadyExists).
```

Somehow it tries to create a RequestValidator that already exists. I can get the deploy to work when I rename the validator just before I deploy using `aws apigateway update-request-validator --rest-api-id {apiID} --request-validator-id {validatorId} --patch-operations op=replace,path=/name,value=some-other-name`

Is there a way to turn off the request validators? I don't really need those and have no idea why they're turned on here but not on some of my other APIs

### Service configuration (serverless.yml) content

```yaml
N/A
have not managed to reproduce with a clean configuration
```

### Command name and used flags

sls deploy --verbose

### Command output

```shell
NA
```

### Environment information

```shell
Operating System:          linux
    Node Version:              14.17.2
    Framework Version:         2.62.0 (local)
    Plugin Version:            5.4.7
    SDK Version:               4.3.0
    Components Version:        3.17.1
```

**Comment by pgrzesik** [comment]
In such case the validator name will be unique across service name and stage - are you somehow deploying services with the same name multiple times to the same stage?

**Comment by necojackarc** [comment]
I saw a similar issue today. When I deployed to production, I got the same error but I didn't when I deployed to testing. I mean, I deployed the same code into two stages, then one worked, but the other didn't.

It seemed like Cloudformation created the validator, tried to create the same one again, then failed with the error. It also failed to roll back, so I needed to manually delete the validator (N.B. this was our first deployment with a newer version of Serverless, so we didn't have the validator till today).

Subsequently, I started seeing `"Too Many Requests (Service: ApiGateway, Status Code: 429, Request ID: XXX, Extended Request ID: null)"` on that stage while I managed to deploy to the other stage..., so I decided to just leave it for a while. I felt Cloudformation was working a bit weirdly today.

I just deployed to the stage again, then it worked...!

I don't know if you are in the same situation but what I did is basically,

- Check if there's the validator created (`aws apigateway get-request-validators --rest-api-id XXX`)
- Delete the validator (`aws apigateway delete-request-validator --rest-api-id XXX --request-validator-id XXX`)
- Wait for a couple of hours
- Redeploy

🤯

**Comment by necojackarc** [comment]
@pgrzesik 

I can see only one `CREATE_IN_PROGRESS` before the first `CREATE_FAILED` with "Request Validator name must be unique", so it _looks_ like it failed on the first creation for some reason.

...but I just noticed the first error was caused by our custom validator added to the resources section, not the one to be created by the Serverless framework. However, I'm very sure this custom validator didn't exist until this time.

In subsequent runs, I renamed this custom validator to see if the issue would go away, and what happened was the validator defined by the Serverless framework internally failed to be created with the same error "Request Validator name must be unique".

Basically, I saw this error with two different validators, so that was another reason I suspected AWS was not working properly at that moment.

**Comment by pgrzesik** [comment]
Thanks for clarification @necojackarc - that definitely sounds like something funky was happening under the AWS hood during these deployments. If any of you experience more issues like this, please report it here so we will be able to nail down the potential cause in Framework :+1:

**Comment by HarrisonJackson** [comment]
Just another data point....Ran into this today setting up a new comp for our serverless deployments. Could not work around it until we downgraded. We've got a `resources` repo that deploys our gateway which is shared in a few other repos with just a bunch of endpoints. We could not deploy any of our endpoint repos on `2.65.0`.

`macOS 12.0.1`

Not working:
```
serverless --version
Framework Core: 2.65.0
Plugin: 5.5.1
SDK: 4.3.0
Components: 3.17.2
```

Working:
```
serverless --version
Framework Core: 2.41.2
Plugin: 5.5.1
SDK: 4.3.0
Components: 3.17.2
```

Didn't attempt any other versions since I knew the lower version worked earlier today with the exact same code / config. Anyone stuck here's all I had to do `yarn global add serverless@2.41.2` for now...

We are using `serverless.ts` if that makes a difference, too.

**Comment by pgrzesik** [comment]
Hello @cbeamond - could you share a little bit more about your setup? Any chance you could provide a minimal reproducible case that we could use for debugging?

**Comment by tszpinda** [comment]
Hi, getting the same problem upgraded from 2.46 to 2.66.2. Went through versions and it works fine on 2.52 but not 2.53.
So quite possible it's something to do with https://github.com/serverless/serverless/pull/9793

Looking at the cloud formation stack events it seems that my stack didn't include the validator before, but after upgrade it's being created and it fails. Any idea what to try next?

**Comment by pgrzesik** [comment]
Hello @tszpinda :wave: Would you be able to provide a reproducible example of that issue?

**Comment by tszpinda** [comment]
@pgrzesik can't reproduce with a simple project :/ but stripping down my project to single API and single lambda it still fails when I have:
`          integration: lambda
          request:
            parameters:
              paths:
                clientId: true
                customReportId: true`
if I remove request.paramteres ... it works.

**Comment by JacobMacInnis** [comment]
Also getting same error. Unable to give a reproducible one because it only happens in one of our stages. We deploy to 8 different environments/stages and one of them is stuck with the issue.
we are on _serverless: 2.64.1_ and have been on it for a while now with no issue. 

among other attempts we have tried removing the Name property with no success and removing it completely from the .yml

**Comment by dave-jumptech** [comment]
I saw the same issue today having upgraded from serverless v1 to v2.66.1 - 2 of my stages deployed fine from packages, one failed with the above error. 

Following the suggestion of @necojackarc, I removed the validator and `sls deploy`ed, then all was well. I didn't wait between removal and re-deploy: 
`aws apigateway get-request-validators --rest-api-id XXX`
`aws apigateway delete-request-validator --rest-api-id XXX --request-validator-id XXX`

**Comment by pgrzesik** [comment]
Hello everyone reporting - one additional clarification - are you using shared `restApiId` by any chance?

**Comment by zoellner** [comment]
I don't. All I do is to reference `!Ref ApiGatewayRestApi` in a few places, for example to create a `AWS::ApiGateway::Authorizer`

**Comment by tszpinda** [comment]
Nope, I don't. 

What's also weird, I have multiple services (12) and all of them use the same configuration - and only one of them fails. The diff between them is the service name, rest API paths and parameters.

**Comment by dave-jumptech** [comment]
@pgrzesik We don't.
```yaml
provider:
  apiGateway:
    minimumCompressionSize: 100 
```

We share the ID as an output of the stack, but nothing beyond that:
``` yaml
Outputs:
  ApiGatewayRestApi:
    Value:
      Ref: ApiGatewayRestApi
    Export:
      Name: CoreApiGatewayRestApi-${opt:stage}::Id
```

**Comment by pgrzesik** [comment]
Thanks a lot for extra details - I will try to understand the issue more and potentially reproduce it on my side.

**Comment by pgrzesik** [comment]
Thanks for sharing @dave-jumptech. Would you be able to share a small reproducible example or a little bit more details about your specific config? I've looked into the implementation and request validator's name includes both `service` and `stage` which should make it unique across different services deployed and I cannot reproduce this issue on my side. Any help/extra details would be really appreciated

**Comment by akowalsk** [comment]
I just experienced this going from 2.25.0 to 2.68.0.  I have 10 deployments with 3 stages each and it only happened on 2 of the 30 (two different deployments, one stage of each that wasn't the same between the deployments).  I'm not sure if this info will help, but it seems to have to do with consolidating all the request validators into a single one.

Output of `aws apigateway get-request-validators --rest-api-id XXX` from 2.25.0 deployment:
```
{
    "items": [
        {
            "id": "8kx2oq",
            "name": "api-c-ApiGa-10F9O8Y9NVNXS",
            "validateRequestBody": true,
            "validateRequestParameters": true
        },
        {
            "id": "xtawqn",
            "name": "api-c-ApiGa-1W0QGKQ67GBC6",
            "validateRequestBody": true,
            "validateRequestParameters": true
        },
        {
            "id": "wjahb2",
            "name": "api-c-ApiGa-1F1OZJVKXK8HA",
            "validateRequestBody": true,
            "validateRequestParameters": true
        }
    ]
}
```

Output of `aws apigateway get-request-validators --rest-api-id XXX` from 2.68.0 **successful** deployment:
```
{
    "items": [
        {
            "id": "g6xe1q",
            "name": "api-customer-staging | Validate request body and querystring parameters",
            "validateRequestBody": true,
            "validateRequestParameters": true
        }
    ]
}
```

Output of `aws apigateway get-request-validators --rest-api-id XXX` from 2.68.0 **failed** deployment:
```
{
    "items": [
        {
            "id": "1oy5pt",
            "name": "api-c-ApiGa-X7MCTM03DCKE",
            "validateRequestBody": true,
            "validateRequestParameters": true
        },
        {
            "id": "3ozgni",
            "name": "api-customer-dev | Validate request body and querystring parameters",
            "validateRequestBody": true,
            "validateRequestParameters": true
        },
        {
            "id": "u51q0t",
            "name": "api-c-ApiGa-XVVXK451F2X7",
            "validateRequestBody": true,
            "validateRequestParameters": true
        },
        {
            "id": "vk93bz",
            "name": "api-c-ApiGa-1U228M5ILIL5C",
            "validateRequestBody": true,
            "validateRequestParameters": true
        }
    ]
}
```

I don't have a small reproducible example to share but I was hoping the above info might be helpful.  I checked the CloudFormation deployment logs of the previous deployment before the error and there was no request validator with the "logical ID" of "ApiGateway<service name>RequestValidator", so it's unclear how it's getting there (and why it isn't getting deleted when the deployment rolls back).

I can also confirm that removing the request validator named per `<service>-<stage> | Validate request body and querystring parameters` per @necojackarc allows me to deploy successfully afterwards.

**Comment by dave-jumptech** [comment]
@pgrzesik I think the short answer is that I can't recreate the issue in another app, in fact I'm not sure I can recreate the issue in the app where I found it now that it's fixed! What other info would be useful for you?

**Comment by pgrzesik** [comment]
Hello @akowalsk - are you sharing API Gateway between different services?

@dave-jumptech I'm not sure what else could be useful to be honest - I was trying to reproduce by issuing multiple deployments but I cannot reproduce with a small test case. If you experience the error again, could you please try to trace to which CF stack belongs the validator that already exists? Maybe that way you will be able to trace how it was created.

**Comment by jon-flowers** [comment]
> * Check if there's the validator created (`aws apigateway get-request-validators --rest-api-id XXX`)
> * Delete the validator (`aws apigateway delete-request-validator --rest-api-id XXX --request-validator-id XXX`)
> * Wait for a couple of hours
> * Redeploy
> 
> 🤯

I hit this repeatedly today in one of my serverless applications that has been around for a couple years.  It generally appears that for some reason Serverless "loses track" of its previously deployed RequestValidator (notably, it's not any one that will be defined in the serverless.yaml, but the "under the hood" one it creates itself).  It then fails to deploy because it tries to create a new one that has the exact same name as the one that already exists.

This solution worked for me; reposting because it didn't have many reactions and it's clear many people are hitting this problem. **If you use the CLI to delete the existing validator and then run a deployment again, it should work.**  As mentioned above, you have to delete the one that serverless creates automatically; for instance, I had two validators.  One specifically created in my YAML, the other with the name "$service-$stage | Validate request body and querystring parameters". I had to delete the latter with the CLI for my deployment to work

What I don't know yet is if I will hit this problem again on my next deploy / update, but if I do I will let you know.

**Comment by igoralves1** [comment]
I had the same issue today and I had no clue what is taht about. 
`An error occurred: XXXXRequestValidator - Resource handler returned message: "Request Validator name must be unique (Service: ...`
Just waited some minutes and deployed again. Seems that the CloudFormation was busy.

**Comment by gabroe** [comment]
Thanks to @jon-flowers we followed his advice and it worked out, just fyi to get the list of validators I used this CLI

`aws apigateway get-request-validators --rest-api-id XXXX`

```
{
    "items": [
        {
            "id": "YYYY",
            "name": "serverless-api-dev | Validate request body and querystring parameters",
            "validateRequestBody": true,
            "validateRequestParameters": true
        }
    ]
}
```

then, you will see the list and use the ID to delete it 


`aws apigateway delete-request-validator --rest-api-id XXXX --request-validator-id  YYYY`

**Comment by mbloch1986** [comment]
I've encountered this issue today as well and a look into Cloudtrail shows an error in the mishandling of `TooManyRequestsException` errors internally ? Looking in details I see this...

Multiple `CreateRequestValidator` calls for each individual request validator ~15 within 2 seconds ... after a few seconds CloudFormation receives this error `TooManyRequestsException`. Based on the information in `userIdentity.sessionContext.attributes.creationDate` it's for one of the first requests ... after that error i see a new `CreateRequestValidator` API call for one of the request validator which were triggered initially. Though based on the creationDate in `userIdentity.sessionContext.attributes.creationDate` of the `TooManyRequestsException` message, the `CreateRequestValidator` call is not for the method for which I have received the `TooManyRequestsException` error.

I assume that CloudFormation re triggers a `CreateRequestValidator` request but maybe for the wrong resource ?
